### PR TITLE
style: swap tslint for eslint typescript

### PR DESCRIPTION
**Pull request by ChristianMurphy**
use eslint rules, eslint typescript rules, import rules, and node js rules.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  "parser": "@typescript-eslint/parser",
+  "parserOptions": {
+    "ecmaVersion": 2020,
+    "sourceType": "module",
+    "project": "./tsconfig.json"
+  },
+  "extends": [
+    "eslint:recommended",
+    "plugin:@typescript-eslint/eslint-recommended",
+    "plugin:@typescript-eslint/recommended",
```
Comment (ChristianMurphy) [inline .eslintrc.json]: this checks typescript best practices
https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#supported-rules

File: .eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
+{
+  "parser": "@typescript-eslint/parser",
+  "parserOptions": {
+    "ecmaVersion": 2020,
+    "sourceType": "module",
+    "project": "./tsconfig.json"
+  },
+  "extends": [
+    "eslint:recommended",
```
Comment (ChristianMurphy) [inline .eslintrc.json]: this checks javascript best practices
https://eslint.org/docs/rules/

File: .eslintrc.json [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    "plugin:@typescript-eslint/eslint-recommended",
+    "plugin:@typescript-eslint/recommended",
+    "plugin:node/recommended",
+    "plugin:import/errors",
+    "plugin:import/warnings",
+    "plugin:import/typescript",
+    "plugin:prettier/recommended"
```
Comment (ChristianMurphy) [inline .eslintrc.json]: this applies the prettier code formatter
https://github.com/prettier/prettier

I left it with the defaults, feel free to configure or replace with another code style
https://prettier.io/docs/en/options.html

File: .eslintrc.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    "project": "./tsconfig.json"
+  },
+  "extends": [
+    "eslint:recommended",
+    "plugin:@typescript-eslint/eslint-recommended",
+    "plugin:@typescript-eslint/recommended",
+    "plugin:node/recommended",
+    "plugin:import/errors",
+    "plugin:import/warnings",
+    "plugin:import/typescript",
```
Comment (ChristianMurphy) [inline .eslintrc.json]: This checks es module best practices
https://github.com/benmosher/eslint-plugin-import#eslint-plugin-import

File: .eslintrc.json [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  "parserOptions": {
+    "ecmaVersion": 2020,
+    "sourceType": "module",
+    "project": "./tsconfig.json"
+  },
+  "extends": [
+    "eslint:recommended",
+    "plugin:@typescript-eslint/eslint-recommended",
+    "plugin:@typescript-eslint/recommended",
+    "plugin:node/recommended",
```
Comment (ChristianMurphy) [inline .eslintrc.json]: This checks Node js best practices
https://github.com/mysticatea/eslint-plugin-node#eslint-plugin-node

File: .eslintrc.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    "plugin:import/errors",
+    "plugin:import/warnings",
+    "plugin:import/typescript",
+    "plugin:prettier/recommended"
+  ],
+  "rules": {
+    "node/no-unsupported-features/es-syntax": [
+      "error",
+      {
+        "ignores": ["modules"]
```
Comment (ChristianMurphy) [inline .eslintrc.json]: Typescript will convert ES Modules to Common JS so Node.js can use them

File: package.json [hunk 7]
```diff
@@ -16,7 +16,9 @@
   "scripts": {
     "build": "tsc",
     "watch": "tsc -w",
-    "test": "ts-node node_modules/jasmine/bin/jasmine"
+    "test": "ts-node node_modules/jasmine/bin/jasmine",
+    "lint": "eslint src --ext ts",
```
Comment (ChristianMurphy) [inline package.json]: runs the linter

File: package.json [hunk 8]
```diff
@@ -16,7 +16,9 @@
...
+    "format": "eslint src --ext ts --fix"
```
Comment (ChristianMurphy) [inline package.json]: fixes lint errors automatically

File: .eslintrc.json [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    "@typescript-eslint/explicit-function-return-type": "off",
+    "import/named": "off"
+  },
+  "settings": {
+    "import/parsers": { "@typescript-eslint/parser": [".ts"] },
+    "node": {
+      "convertPath": {
+        "src/**/*.ts": ["^src/(.+?)\\.ts$", "dist/$1.js"]
+      },
+      "tryExtensions": [".js", ".ts"]
```
Comment (ChristianMurphy) [inline .eslintrc.json]: let's node plugin find typescript extension'ed files

File: .eslintrc.json [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      }
+    ],
+    "@typescript-eslint/explicit-function-return-type": "off",
+    "import/named": "off"
+  },
+  "settings": {
+    "import/parsers": { "@typescript-eslint/parser": [".ts"] },
+    "node": {
+      "convertPath": {
+        "src/**/*.ts": ["^src/(.+?)\\.ts$", "dist/$1.js"]
```
Comment (ChristianMurphy) [inline .eslintrc.json]: let's the node plugin know that transpilation is happening, and the `dist` folder is constructed from the `src` folder

File: .eslintrc.json [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      "error",
+      {
+        "ignores": ["modules"]
+      }
+    ],
+    "@typescript-eslint/explicit-function-return-type": "off",
+    "import/named": "off"
+  },
+  "settings": {
+    "import/parsers": { "@typescript-eslint/parser": [".ts"] },
```
Comment (ChristianMurphy) [inline .eslintrc.json]: lets the import module find typescript extension'ed files

File: package.json [hunk 12]
```diff
@@ -4,7 +4,7 @@
   "description": "Accepts a directory full of json-schema files and produces associated validators and TypeScript interfaces. Great for validating server side with NodeJS and improving TypeScript client side dev.",
   "main": "dist/index.js",
   "files": [
-    "dist/index.js"
+    "dist"
```
Comment (ChristianMurphy) [inline package.json]: this code is not bundled, all the files in the dist folder are needed when packaging for NPM

File: package.json [hunk 13]
```diff
@@ -26,24 +28,36 @@
     "ajv-pack": "^0.3.1",
     "fs-extra": "^8.1.0",
     "json-schema-to-typescript": "^7.1.0",
+    "readdirp": "^3.1.2",
```
Comment (ChristianMurphy) [inline package.json]: this is used by `get-all-files.ts` script

File: src/index.ts [hunk 14]
```diff
@@ -1,49 +1,55 @@
 #!/usr/bin/env node
-import fs from 'fs-extra';
-import path from 'path';
-import yargs from 'yargs';
-import { compileFromFile } from 'json-schema-to-typescript';
-import getAllFiles from './scripts/get-all-files';
-import generateJsonSchemaValidators from './scripts/generate-json-schema-validators.js';
-
-interface Options {
```
Comment (ChristianMurphy) [inline src/index.ts]: interface was an artifact left over from strict typescript refactor that turned out not to be needed, removed.

**Comment by gitsome** [comment]
Awesome, @ChristianMurphy this all looks great. Thank you! Merging away.... and using a lot of what you did here on other projects now :-)

**Pull request merged**